pkg/relayer/session/sessiontree.go

```diff
@@ -0,0 +1,190 @@
+	}
+
+	return sessionTree, nil
+}
+
+// GetSession returns the session corresponding to the SMST.
+func (st *sessionTree) GetSession() *sessiontypes.Session {
+	return st.session
+}
+
+// Update is a wrapper for the SMST's Update function. It updates the SMST with the given key, value, and weight.
+// This function should be called by the Miner when a Relay has been successfully served.
+// It returns an error if the SMST has been flushed to disk which indicates that updates are no longer allowed.
+func (st *sessionTree) Update(key, value []byte, weight uint64) error {
+	st.sessionMu.Lock()
+	defer st.sessionMu.Unlock()
+
+	if st.claimedRoot != nil {
+		return ErrSessionTreeClosed
+	}
+
+	return st.tree.Update(key, value, weight)
+}
+
+// ProveClosest is a wrapper for the SMST's ProveClosest function. It returns a proof for the given path.
+// This function should be called when a session has been claimed and needs to be proven.
+// If the proof has already been generated, it returns the cached proof.
+// It returns an error if the SMST has not been flushed yet (the claim has not been generated)
+func (st *sessionTree) ProveClosest(path []byte) (proof *smt.SparseMerkleClosestProof, err error) {
+	st.sessionMu.Lock()
+	defer st.sessionMu.Unlock()
+
+	// A claim need to be generated before a proof can be generated.
+	if st.claimedRoot == nil {
+		return nil, ErrSessionTreeNotClosed
+	}
+
+	// If the proof has already been generated, return the cached proof.
+	if st.proof != nil {
+		return st.proof, nil
+	}
+
+	// Restore the KVStore from disk since it has been closed after the claim has been generated.
+	st.treeStore, err = smt.NewKVStore(st.storePath)
+	if err != nil {
+		return nil, err
+	}
+
+	st.tree = smt.ImportSparseMerkleSumTree(st.treeStore, sha256.New(), st.claimedRoot, smt.WithValueHasher(nil))
+
+	// Generate the proof and cache it along with the path for which it was generated.
+	st.proof, err = st.tree.ProveClosest(path)
+	st.proofPath = path
+
+	return st.proof, err
+}
+
+// Flush gets the root hash of the SMST needed for submitting the claim; then commits the entire tree to disk
+// and stops the KVStore.
+// It should be called before submitting the claim on-chain. This function frees up the in-memory resources.
+// If the SMST has already been flushed to disk, it returns the cached root hash.
+func (st *sessionTree) Flush() (SMSTRoot []byte, err error) {
+	st.sessionMu.Lock()
+	defer st.sessionMu.Unlock()
+
+	// We already have the root hash, return it.
+	if st.claimedRoot != nil {
+		return st.claimedRoot, nil
+	}
+
+	st.claimedRoot = st.tree.Root()
+
+	// Commit the tree to disk
+	if err := st.tree.Commit(); err != nil {
+		return nil, err
+	}
+
+	// Stop the KVStore
+	if err := st.treeStore.Stop(); err != nil {
+		return nil, err
+	}
+
+	st.treeStore = nil
+	st.tree = nil
+
+	return st.claimedRoot, nil
+}
+
+// Delete deletes the SMST from the KVStore and removes the sessionTree from the sessionManager.
```

Comment:
You should update the comment tha tit also deletes the kvstore from disk and add a WARNING label.